architecture/crypto.rst

```diff
@@ -191,9 +191,51 @@ GP requirements tested and covered by the OP-TEE sanity test suite
 supported - i.e: the SE05x does not implement all RSA key sizes - we opted for
 disabling those particular tests at build time rather than letting them fail.
 
+NXP SE05X Family of Secure Elements
+***********************************
+
+This family of I2C bus devices are supported through the se050 cryptographic driver
+located at `core/drivers/crypto/se050`_. Before the REE boots, the session with
+the device is established using one of the OP-TEE supported I2C platform device
+drivers. Once the REE is up, the cryptographic driver can be configured to use
+the I2C driver in the REE (via RPC service) or continue using the one in OP-TEE.
+
+Unless the Secure Element owns the I2C bus (no other elements on the bus, no
+runtime-PM and so forth), it is recommended to route all traffic via the Normal
+World. Initial communication with the device is not data intensive and therefore
+slow I2C drivers - perhaps those not using DMA channels - do not represent much
+of a performance drag; the situation changes once clients start hammering the
+device.
+
+If using the REE for I2C transfers, it is also recommended to configure the
+driver so that the Secure Channel Protocol 03 is enabled prior to exiting the
+Secure World; this way all communication between the processor and the secure
+element is encrypted and MAC authenticated.
+
+Aside of the secure element integration as an OP-TEE cryptographic driver,
+OP-TEE  also presents an Application Protocol Data Units (APDU) interface to
+users via its  OP-TEE client.
+
+.. figure:: ../images/crypto/drivers/se050_apdu_pta_interface.png
+    :figclass: align-center
+
+    Access to the Secure Element from libseetec and the APDU PTA.
+
+Using this interface, priviledged applications can control the Secure Element to
+inject or delete keys or certificates, encrypt, decrypt, sign and verify data
+and so forth. An application implementing a subset of those functions can be
+seen in this Foundries.io `repository`_. 
+
+This reference code wont be fully functional in OP-TEE upstream as the patches
```

Comment:
This passage is confusing. The purpose of this git is to document upstream OP-TEE.

Reply:
These commands will still work:

```
fio-se05x-cli --delete-objects all
fio-se05x-cli --delete-objects 0x67981234 
fio-se05x-cli --show-cert 0xf0000123 
```

Reply:
This was the OP-TEE change required to support the additional functionality 
https://github.com/foundriesio/optee_os/commit/67bbf1f816aaa6b6f0a5c92d1da591f3281a3723

Reply:
Why is it that you can do this in a standard way? Shortcomings in PKCS#11 or TEE Internal Core API?

Reply:
I did not find a way of addressing an existing key that already exists in the HSM using PCSK#11 (not so much the TEE core); my work around was to appended the key identifier to the **SE_** prefix in the LABEL and use the label to get the id in the TA.

Reply:
I would have rather extended the private key definition with a field to support HSM handles but there were so many changes at the time that I opted for the minimum amount of changes (ie, the watermark of handles). Perhaps we should revisit it.

Reply:
I don't have an opinion on this at the moment. Perhaps @etienne-lms or @vesajaaskelainen has some thoughts on this.

Reply:
Your issue @ldts is that at key creation, you don't have a generic PKCS#11 interface to tell that the key must land in a very specific place of the token.

From GP side, GP 1.3 allows implementation defined object usage flags (mask 0xFF000000), one could reflect how the key should be used, e.g. being bound to a external device. This way, i guess a bit flag is not enough, there may need a device identifier or like.

But that won't resolve how the info can be passed through the PKCS#11 API. Looking again in the spec, maybe CKA_HW_FEATURE_TYPE / CK_HW_FEATURE_TYPE could help. Note there's only place for vendor defined type IDs here.

Reply:
> Your issue @ldts is that at key creation, you don't have a generic PKCS#11 interface to tell that the key must land in a very specific place of the token.
> 
> From GP side, GP 1.3 allows implementation defined object usage flags (mask 0xFF000000), one could reflect how the key should be used, e.g. being bound to a external device. This way, i guess a bit flag is not enough, there may need a device identifier or like.
> 
> But that won't resolve how the info can be passed through the PKCS#11 API. Looking again in the spec, maybe CKA_HW_FEATURE_TYPE / CK_HW_FEATURE_TYPE could help. Note there's only place for vendor defined type IDs here.

Right. and that is the kind of conversation we should have had but unfortunately we were not ready to have in OP-TEE back then. Maybe PKCS#11 maintainers could discuss how to support secure elements - I'd be happy to contribute with use cases. 

In any case, what do I do with this PR.  It seems to be taking everyones time that maybe is better invested somewhere else. I could close it and we can revisit all next year if you want.

Reply:
Main problem with external crypto key material provider is that how to make sure that keys managed within PKCS#11 TA can only be operated within the TA. Eg. you cannot operate with the key behind the back of PKCS#11 TA's access control restrictions.

When PKCS#11 TA owns the key material and they are stored within the TA's private secure storage they are only accessible by the TA itself and then all access control restrictions apply.

Also I recall that you were planning to import other keys created elsewhere and possibly operated beforehand into the PKCS#11 TA's key space.

Reply:
Um, why not? is that part of some requirement?

The implementation we use - I posted some links above - achieve this by applying the logic on the cryptographic driver key generation interface: the PKCS#11 TA will store not a private key but a handle to the key securely stored in the HSM: PKCS#11 operations requiring to use the key have to be supported by the HSM cryptographic driver, that is all. This key can be used outside PKCS#11 TA of course  since clients will just need to know the handle...all operations with such a key is done always inside the HSM, the key never leaks.

Reply:
I updated this section on the recent push.